runtime/src/accounts.rs

```diff
@@ -550,49 +550,93 @@ impl Accounts {
-        self.store_credit_only_credits(w_credit_only_account_locks.drain(), ancestors, fork);
+        self.store_credit_only_credits_and_rents(
+            w_credit_only_account_locks.drain(),
+            rent_collector,
+            ancestors,
+            fork,
+        )
     }
 
-    fn store_credit_only_credits<I>(
+    fn store_credit_only_credits_and_rents<I>(
         &self,
         credit_only_account_locks: I,
+        rent_collector: &RentCollector,
         ancestors: &HashMap<Fork, usize>,
         fork: Fork,
-    ) where
+    ) -> (u64, u64)
+    where
         I: IntoIterator<Item = (Pubkey, CreditOnlyLock)>,
     {
-        for (pubkey, lock) in credit_only_account_locks {
-            let lock_count = *lock.lock_count.lock().unwrap();
-            if lock_count != 0 {
-                warn!(
-                    "dropping credit-only lock on {}, still has {} locks",
-                    pubkey, lock_count
-                );
-            }
-            let credit = lock.credits.load(Ordering::Relaxed);
-            if credit > 0 {
-                let mut account = self
-                    .load_slow(ancestors, &pubkey)
-                    .map(|(account, _)| account)
-                    .unwrap_or_default();
+        let mut accounts: Vec<(Pubkey, Account)> = vec![];
+        let mut total_rent_collected = 0;
+        let mut total_credit_credited = 0;
+
+        {
+            let accounts_index = self.accounts_db.accounts_index.read().unwrap();
+            let storage = self.accounts_db.storage.read().unwrap();
+            for (pubkey, lock) in credit_only_account_locks {
+                let lock_count = *lock.lock_count.lock().unwrap();
+                if lock_count != 0 {
+                    warn!(
+                        "dropping credit-only lock on {}, still has {} locks",
+                        pubkey, lock_count
+                    );
+                }
+                let credit = lock.credits.load(Ordering::Relaxed);
+
+                let (mut account, _) =
+                    AccountsDB::load(&storage, ancestors, &accounts_index, &pubkey)
+                        .unwrap_or_default();
                 account.lamports += credit;
-                self.store_slow(fork, &pubkey, &account);
+                total_credit_credited += credit;
+
+                // We don't want newly created credit account to be saved if it cannot even
+                // pay for current cycle of rent.
+                if let Some((account, rent)) = rent_collector.update(account) {
+                    total_rent_collected += rent;
+                    accounts.push((pubkey, account));
+                }
             }
         }
+
+        let account_to_store: Vec<(&Pubkey, &Account)> = accounts
+            .iter()
+            .map(|(key, account)| (key, account))
+            .collect();
+
+        self.accounts_db.store(fork, &account_to_store);
+
+        (total_credit_credited, total_rent_collected)
```

Comment:
looks to me like no one cares about total_credits_credited, why sum and return it?

Reply:
fixed, only returning total_rent_collected.